File: packages/ts/src/domain/tool.ts
import type { ToolCallId } from "./identifiers";

/** Risk classification that drives whether a tool call needs human approval. */
export const RiskTier = {
  /** Read-only; runs automatically. */
  Safe: "safe",
  /** Side-effecting; gated behind human approval. */
  Sensitive: "sensitive",
} as const;
export type RiskTier = (typeof RiskTier)[keyof typeof RiskTier];

/** A JSON Schema object describing a tool's parameters. */
export type JsonSchema = Readonly<Record<string, unknown>>;

/** What the model is shown about a tool. Carries no executable behavior. */
export interface ToolDefinition {
  readonly name: string;
  readonly description: string;
  readonly parameters: JsonSchema;
  readonly risk: RiskTier;
}

/** The model's request to invoke a tool. */
export interface ToolCall {
  readonly id: ToolCallId;
  readonly name: string;
  readonly args: Readonly<Record<string, unknown>>;
}

/** Executes a tool's side effect. The only place an effect happens. */
export type ToolHandler = (args: Readonly<Record<string, unknown>>) => Promise<unknown>;

/** A tool definition bound to its handler. The handler is never sent to the model. */
export interface Tool extends ToolDefinition {
  readonly handler: ToolHandler;
}

File: packages/py/src/airlock/domain/tool.py
"""Tools: what the agent can do, and how risky each one is."""

from collections.abc import Awaitable, Callable, Mapping
from dataclasses import dataclass
from enum import StrEnum

from pydantic import BaseModel, ConfigDict

from airlock.domain.identifiers import ToolCallId


class RiskTier(StrEnum):
    """Drives whether a tool call needs human approval."""

    SAFE = "safe"
    SENSITIVE = "sensitive"


class ToolDefinition(BaseModel):
    """What the model is shown about a tool. Carries no executable behavior."""

    model_config = ConfigDict(frozen=True)

    name: str
    description: str
    parameters: dict[str, object]
    risk: RiskTier


class ToolCall(BaseModel):
    """The model's request to invoke a tool."""

    model_config = ConfigDict(frozen=True)

    id: ToolCallId
    name: str
    args: dict[str, object]


ToolHandler = Callable[[Mapping[str, object]], Awaitable[object]]
"""Executes a tool's side effect. The only place an effect happens."""


@dataclass(frozen=True)
class Tool:
    """A tool definition bound to its handler; the handler never reaches the model."""

    name: str
    description: str
    parameters: dict[str, object]
    risk: RiskTier
    handler: ToolHandler

    def definition(self) -> ToolDefinition:
        """The model-visible definition, without the handler."""
        return ToolDefinition(
            name=self.name,
            description=self.description,
            parameters=self.parameters,
            risk=self.risk,
        )
